src/opuntia.erl

```diff
@@ -0,0 +1,53 @@
+%% @doc `opuntia', traffic shapers for Erlang and Elixir
+%%
+%% This module implements the token bucket traffic shaping algorithm.
+%% The time unit of measurement is millisecond, as this is the unit receives and timers use in the BEAM.
+%% @end
+-module(opuntia).
+
+-export([new/1, update/2]).
+
+-include("opuntia.hrl").
+
+-type timestamp() :: number().
+-type tokens() :: non_neg_integer().
+-type delay() :: non_neg_integer().
+-type rate() :: non_neg_integer().
+-type shaper() :: #token_bucket{} | none.
+
+-export_type([tokens/0, timestamp/0, delay/0, rate/0, shaper/0]).
+
+-spec new(rate()) -> shaper().
+new(0) ->
+    none;
+new(MaxRatePerMs) ->
+    #token_bucket{rate = MaxRatePerMs,
+                  available_tokens = MaxRatePerMs,
+                  last_update = erlang:monotonic_time(millisecond)}.
+
+%% @doc Update shaper and return possible waiting time.
+%%
+%% This function takes the current shaper state, and the number of tokens that have been consumed,
+%% and returns a tuple containing the new shaper state, and a possibly non-zero number of
+%% unit times to wait if more tokens that the shaper allows were consumed.
+-spec update(shaper(), tokens()) -> {shaper(), rate()}.
+update(none, _Size) ->
+    {none, 0};
+update(#token_bucket{rate = MaxRatePerMs,
+                     available_tokens = LastAvailableTokens,
+                     last_update = LastUpdate} = Shaper, TokensNowUsed) ->
+    Now = erlang:monotonic_time(millisecond),
+    % How much we might have recovered since last time
+    TimeSinceLastUpdate = Now - LastUpdate,
+    PossibleTokenGrowth = round(MaxRatePerMs * TimeSinceLastUpdate),
+    % Available plus recovered cannot grow higher than the actual rate limit
+    ExactlyAvailableNow = min(MaxRatePerMs, LastAvailableTokens + PossibleTokenGrowth),
+    % Now check how many tokens are available by substracting how many where used,
+    % and how many where overused
+    TokensAvailable = max(0, ExactlyAvailableNow - TokensNowUsed),
+    TokensOverused = max(0, TokensNowUsed - ExactlyAvailableNow),
+    MaybeDelay = TokensOverused / MaxRatePerMs,
+    RoundedDelay = floor(MaybeDelay) + 1,
+    NewShaper = Shaper#token_bucket{available_tokens = TokensAvailable,
+                                    last_update = Now + MaybeDelay},
```

Comment:
The point is that in the bucket I want to keep track of the exact times without rounding errors, so if in this case I asked to wait 0.7ms more than I should have, the bucket knows this and in the next request it will request to wait 0.3ms less, compensating.
But in the returned delay I wanted to give integers and not floats so that they can be given to the corresponding timeout directly.

Reply:
Ok, it all depends on what we do with the `+1` in line. If we replace it with a fair rounding (not always adding a systematic error), your assumptions will be ok. The only think I would point out is that we are losing precision here, because float is only very accurate around zero, while timestamps are very far from zero. I tested it and we still have 4 digits of precision:

```
36> Now = erlang:monotonic_time(millisecond).
-576458575501
37> float(Now) + 0.00001.
-576458575501.0
38> float(Now) + 0.0001.
-576458575500.9999
```